TickSpec/ServiceProvider.fs

```diff
@@ -3,21 +3,94 @@
 open System
 open System.Collections.Generic
 open System.Diagnostics
+open System.Reflection
 
+/// Provides an instance provider for tests
+type IInstanceProvider =
+    inherit IDisposable
+
+    /// Resolves an instance of desired type
+    abstract member Resolve : Type -> obj
+
+    /// Registers an instance for a type (if there is already a registered instance, it will be replaced)
+    abstract member RegisterInstance : Type -> obj -> unit
+
+[<Sealed>]
 /// Creates instance service provider
-type ServiceProvider () =
-    /// Type instances constructed for invoked steps
+type ServiceProvider () as self =
+    /// Type instances for invoked steps
     let instances = Dictionary<_,_>()
+
+    /// Resolves an instance for a specified type (and remembering the stack of types being resolved)
+    let rec resolveInstance (t:Type) (typeStack: Type list) =
+        let alreadyRequested =
+            typeStack
+            |> List.tryFind (fun x -> x = t)
+
+        match alreadyRequested with
+        | Some _ -> raise (InvalidOperationException(sprintf "Circular dependency found when resolving type %O" t))
+        | None -> ()
+
+        match t with
+        | t when t = typeof<IInstanceProvider> -> self :> obj

```

Comment:
in line with the [ISP](https://en.wikipedia.org/wiki/Interface_segregation_principle) it would be good to be able to indicate one only consumes and does not provide by asking for IServiceProvider ?

Having said that, this will probably complicate the impl a bit here

Reply:
supporting `fun unit -> 't` and `fun 't -> unit` (maybe that gets messy) would allow one to bind to an F# type:-

```type MyFeature(resolve, register) =
    let x : DogBowl = resolve()
    let doStuff () =
        register <| SmallDog()
```

that might get us out of the interface game?
Or e.g. [delegate factories](https://autofac.readthedocs.io/en/latest/advanced/delegate-factories.html) can allow us to pass args to ctors.
Why am I mentioning this?

I'm concerned that having a very specific interface and semantics in this container will make it much harder to ever support other semantics. Thinking....

Reply:
So, you prefer to have two separate interfaces for registration and resolving and you would not be able to ask for the complete container, just for one of the two interfaces? I can do that, but I do not understand the reason. Regarding ISP, both methods are for container manipulation, storing and retrieving so from this principle it seems ok to me... Again, I have no problem doing it, although I am not 100% sure, that it is worth it.

Regarding the two methods - they look nice and I can imagine, that we may want to add it as an addition. However, it does not look to be right in terms of DI, because what type you use when you call register in that case? SmallDog, Dog or IDog? I think that it is good to be explicit in this case.

Reply:
Sorry, I misformatted the example and omitted the `type` definition.
The container should be substitutable and is not the interesting bit. For the bulk of what TickSpec wants, IServiceProvider is sufficient.
If we build a way for return values to be registered, then we need a way for that to work.
But in general an author of a Feature type should be able to ask for either 
- 80% of the time (ideally via ctor args), values like normal DI, by supplying params as opposed to explicit calls
- 5 % of the time registering stuff to share with other steps
- 15% of the time interacting in fancy ways - people in that space will probably be best off using Autofac or their favorite container.
Baking an interface which does the register and the resolve doesn't really help anyone in the above categories as I see it. Does this make any sense ?

Re whether the signatures are generic or have an explcit type arg, for resolve that gets very ugly very fast. Also if a container ha preregistered stuff, you dont want to do explicit calls
```
type MyFeature(kennel : Kennel, resolve, register) =
    let x : DogBowl = resolve()
    let doStuff () =
        register <| SmallDog()
    let letTheDogsOut() =
      kennel.letThemOut ()
```
If we do implicit register of the return value, then you're in the land of having to make sure the types line up anyway.

None of this precludes also having a way to supply a `Type` too, but in general, tpying `typeof<X>` is not something I'd want in normal DI code, so why would I want to be doing that in step defintiions ?

Reply:
The other thing is regarding splitting registering and resolving, many containers (such as Autofac) make a very deliberate contrast between registering and resolving from a CQS perspective (I dont even know how you do runtime registers in Autofac as in general, it bakes a graph, but assume there is a way)

From an ISP perspective I'd argue that you're only very rarely doing both registering and resolving in any given context